File: java-suite/src/main/java/de/fraunhofer/iem/classesAndObjects/classes/inheritance/baseAndDerivedClass/Animal.java
package de.fraunhofer.iem.classesAndObjects.classes.inheritance.baseAndDerivedClass;

/**
 * Non-class that can be a superclass to another class.
 *
 * @author Ranjith Krishnamurthy
 */
public class Animal {
    private String DNA;

    public Animal(String DNA) {
        this.DNA = DNA;
    }

    /**
     * Derived class can override this method since it is a non-final method
     */
    public void makeNoise() {
        System.out.println("This is abstract animal sound!");
    }

    /**
     * Derived class can not override this method since it is a final method
     */
    public final void displayDNA() {
        System.out.println(DNA);
    }
}

File: kotlin-suite/src/main/kotlin/de/fraunhofer/iem/classesAndObjects/classes/inheritance/baseAndDerivedClass/Animal.kt
package de.fraunhofer.iem.classesAndObjects.classes.inheritance.baseAndDerivedClass

/**
 * Non-class that can be a superclass to another class because it is declared as open (In Java bytecode, this class will
 * be non-final class).
 *
 * @author Ranjith Krishnamurthy
 */
open class Animal(private val DNA: String) {
    /**
     * Derived class can override this method since it is a open method (In Java bytecode, its non-final method)
     */
    open fun makeNoise() {
        println("This is abstract animal sound!")
    }

    /**
     * Derived class can not override this method since it is a final method (not open method)
     */
    fun displayDNA() {
        println(DNA)
    }
}